src/peer_transport/handler.ts

```diff
@@ -0,0 +1,151 @@
+import type { IncomingStreamData } from '@libp2p/interface-registrar'
+import { pbStream } from 'it-pb-stream'
+import pDefer, { type DeferredPromise } from 'p-defer'
+import { TimeoutController } from 'timeout-abort-controller'
+import { readCandidatesUntilConnected } from './util.js'
+import * as pb from './pb/index.js'
+import { abortableDuplex } from 'abortable-iterator'
+import { logger } from '@libp2p/logger'
+import type { Stream } from '@libp2p/interface-connection'
+import type { StreamMuxerFactory } from '@libp2p/interface-stream-muxer'
+import { DataChannelMuxerFactory } from '../muxer.js'
+
+const DEFAULT_TIMEOUT = 30 * 1000
+
+const log = logger('libp2p:webrtc:peer')
+
+export type IncomingStreamOpts = { rtcConfiguration?: RTCConfiguration } & IncomingStreamData
+
+export async function handleIncomingStream ({ rtcConfiguration, stream: rawStream }: IncomingStreamOpts): Promise<[RTCPeerConnection, StreamMuxerFactory]> {
+  const timeoutController = new TimeoutController(DEFAULT_TIMEOUT)
+  const signal = timeoutController.signal
+  const stream = pbStream(abortableDuplex(rawStream, timeoutController.signal)).pb(pb.Message)
+  const pc = new RTCPeerConnection(rtcConfiguration)
+  const muxerFactory = new DataChannelMuxerFactory(pc)
+
+  const connectedPromise: DeferredPromise<void> = pDefer()
+  signal.onabort = () => { connectedPromise.reject() }
+  // candidate callbacks
+  pc.onicecandidate = ({ candidate }) => {
```

Comment:
I thought it was a bug to send an ICE candidate before an answer (there's a test case for that), why are we setting the callback here rather than after we send the answer?

Reply:
Because firefox will start gathering candidates immediately after we create the offer. This is still a valid flow, since the remote will receive an offer followed by candidates.

Reply:
Isn't it racey then? Couldn't this write to the stream before we write our answer?

Reply:
It's not a  problem in the firefox case,  as it seems the candidates for the peer that generates the answer are gathered after the local description is set, i.e the answer is generated and set. I can add a deferred promise to alleviate the race  condition.

Reply:
Fixed now